### PR TITLE
[AIRFLOW-3993] Change SalesforceHook and add tests

### PR DESCRIPTION
**Pull request by feluelle**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3993
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

- add tests
- refactor code
- update docs
- change sign_in to get_conn
- add salesforce to devel_all packages
- add note to UPDATING.md

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

File: airflow/contrib/hooks/salesforce_hook.py [hunk 1]
```diff
@@ -18,127 +18,108 @@
...
-This module contains a Salesforce Hook
-which allows you to connect to your Salesforce instance,
-retrieve data from it, and write that data to a file
-for other uses.
+This module contains a Salesforce Hook which allows you to connect to your Salesforce instance,
+retrieve data from it, and write that data to a file for other uses.
 
-NOTE:   this hook also relies on the simple_salesforce package:
-        https://github.com/simple-salesforce/simple-salesforce
+NOTE: this hook also relies on the simple_salesforce package:
```
Comment (mik-laj) [inline airflow/contrib/hooks/salesforce_hook.py]:
```suggestion
.. note:: this hook also relies on the simple_salesforce package:
```
Reference: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-note

File: airflow/contrib/hooks/salesforce_hook.py [hunk 2]
```diff
@@ -18,127 +18,108 @@
...
         self.conn_id = conn_id
-        self._args = args
-        self._kwargs = kwargs
-
-        # get the connection parameters
-        self.connection = self.get_connection(conn_id)
-        self.extras = self.connection.extra_dejson
+        self.salesforce = None
 
     def sign_in(self):
```
Comment (mik-laj) [inline airflow/contrib/hooks/salesforce_hook.py]:
```suggestion
     def get_conn(self):
```
It is recommended that the `get_conn` method return the API library. Return should be cached.

**Comment by mik-laj** [comment]
Are all your changes backward compatible? If not, you must add a note in the `UPDATING.rst` file.

**Comment by feluelle** [comment]
@mik-laj They should be, yes.

File: airflow/contrib/hooks/salesforce_hook.py (hunk 2 above)
Comment (feluelle) [reply airflow/contrib/hooks/salesforce_hook.py]: Agree. :)

**Comment by feluelle** [comment]
I don't understand why `simple_salesforce` can not be imported. 
See: https://travis-ci.org/apache/airflow/jobs/501970876#L2878

It is listed in the `extras_require`.
See: https://github.com/apache/airflow/blob/master/setup.py#L217
and https://github.com/apache/airflow/blob/master/setup.py#L375

**Comment by mik-laj** [comment]
is it in devel_all?

**Comment by feluelle** [comment]
Ah okay, I missed this line: [devel_ci = devel_all](https://github.com/apache/airflow/blob/master/setup.py#L273). Thank you @mik-laj I will add it.

**Comment by feluelle** [comment]
The failing check is not related to changes of this PR.

**Comment by Fokko** [comment]
@feluelle the k8s tests are a bit flaky. I've restarted the test.

File: airflow/contrib/hooks/salesforce_hook.py [hunk 3]
```diff
@@ -18,127 +18,109 @@
...
 
 
 class SalesforceHook(BaseHook):
-    def __init__(
-            self,
-            conn_id,
-            *args,
-            **kwargs
-    ):
+    def __init__(self, conn_id):
```
Comment (Fokko) [inline airflow/contrib/hooks/salesforce_hook.py:38]: Could you add the `@apply_defaults` decorator here?

**Review by Fokko — COMMENTED** [review]
Love it, much better than the original. Also more ideomatic Airflow becaus of the `get_conn` instead of the `sign_in`.

File: airflow/contrib/hooks/salesforce_hook.py (hunk 3 above)
Comment (feluelle) [reply airflow/contrib/hooks/salesforce_hook.py:38]: Done. :)

Comment (feluelle) [reply airflow/contrib/hooks/salesforce_hook.py:38]: Wait.. Isn't the `@apply_defaults` only necessary for operators? @Fokko  

> def apply_defaults(func):
    """
    Function decorator that Looks for an argument named "default_args", and
    fills the unspecified arguments from it.

Comment (feluelle) [reply airflow/contrib/hooks/salesforce_hook.py:38]: I reverted it. Please let me know if I am wrong and I will add it back in.

Comment (mik-laj) [reply airflow/contrib/hooks/salesforce_hook.py:38]: this is only for operators. it makes it easier to create DAGs

**Comment by feluelle** [comment]
@Fokko May you restart the failed test? :)

**Comment by feng-tao** [comment]
@feluelle , could you rebase your pr with master?

**Comment by feluelle** [comment]
> @feluelle , could you rebase your pr with master?

Done 👍 @feng-tao

File: airflow/contrib/hooks/salesforce_hook.py [hunk 4]
```diff
@@ -18,127 +18,110 @@
...
-        Create new connection to Salesforce
-        and allows you to pull data out of SFDC and save it to a file.
+        Create new connection to Salesforce and allows you to pull data out of SFDC and save it to a file.
+
+        You can then use that file with other Airflow operators to move the data into another data source.
 
-        You can then use that file with other
-        Airflow operators to move the data into another data source
+        :param conn_id: the name of the connection that has the parameters we need to connect to Salesforce.
+                        The connection should be type `http` and include a user's security token
```
Comment (zhongjiajie) [inline airflow/contrib/hooks/salesforce_hook.py]: This is indented too far for docstring rendering? see https://github.com/apache/airflow/pull/4895#discussion_r268388705

File: airflow/contrib/hooks/salesforce_hook.py [hunk 5]
```diff
@@ -18,127 +18,110 @@
...
-        self.connection = self.get_connection(conn_id)
-        self.extras = self.connection.extra_dejson
-
-    def sign_in(self):
+    def get_conn(self):
         """
-        Sign into Salesforce.
-
-        If we have already signed it, this will just return the original object
+        Sign into Salesforce, only If we are not already signed in.
```
Comment (zhongjiajie) [inline airflow/contrib/hooks/salesforce_hook.py]:
```suggestion
        Sign into Salesforce, only if we are not already signed in.
```
or you just want to point out this?

File: airflow/contrib/hooks/salesforce_hook.py [hunk 6]
```diff
@@ -174,93 +165,83 @@ def _to_timestamp(cls, col):
...
-                                    False if you want them to be left in the
-                                    same format as they were in Salesforce.
-                                    Leaving the value as False will result
-                                    in datetimes being strings.
+        :param query_results: the results from a SQL query
+        :type query_results: list of dict
+        :param filename: the name of the file where the data should be dumped to
+        :type filename: str
+        :param fmt: the format you want the output in.
+                    *Default:* csv.
```
Comment (zhongjiajie) [inline airflow/contrib/hooks/salesforce_hook.py]: indented

Comment (zhongjiajie) [reply airflow/contrib/hooks/salesforce_hook.py]: and below.

File: airflow/contrib/hooks/salesforce_hook.py [hunk 7]
```diff
@@ -303,18 +281,12 @@ def write_object_to_file(
...
             possible_strings = df.columns[df.dtypes == "object"]
-            df[possible_strings] = df[possible_strings].apply(
-                lambda x: x.str.replace("\r\n", "")
-            )
-            df[possible_strings] = df[possible_strings].apply(
-                lambda x: x.str.replace("\n", "")
-            )
-
+            df[possible_strings] = df[possible_strings].apply(lambda x: x.str.replace("\r\n", ""))
+            df[possible_strings] = df[possible_strings].apply(lambda x: x.str.replace("\n", ""))
```
Comment (zhongjiajie) [inline airflow/contrib/hooks/salesforce_hook.py]: Could we use single line to remove rather than L288-L289?

**Review by zhongjiajie — COMMENTED** [review]
My 2 cents

File: airflow/contrib/hooks/salesforce_hook.py (hunk 7 above)
Comment (feluelle) [reply airflow/contrib/hooks/salesforce_hook.py]: Yes, thanks for pointing that out. 👍

File: airflow/contrib/hooks/salesforce_hook.py (hunk 4 above)
Comment (feluelle) [reply airflow/contrib/hooks/salesforce_hook.py]: Oh really? Did not know that. I have not checked the docs on mobile. Will fix this.

File: airflow/contrib/hooks/salesforce_hook.py (hunk 7 above)
Comment (zhongjiajie) [reply airflow/contrib/hooks/salesforce_hook.py]: You welcome.

**Comment by feluelle** [comment]
I rebased once again. 
Please @ashb @Fokko @feng-tao @XD-DENG let me know if you have any further comments. Thanks :)

**Review by Fokko — APPROVED** [review]
Looking good @feluelle Thanks!

**Comment by ashb** [comment]
Other than the function rename (which is easy enough for me to add a deprecation shim) is this okay pull in to 1.10.4?

**Comment by Fokko** [comment]
Officially we should deprecate the function first, but I'm fine with pulling it into 1.10.4

**Pull request merged**